Refactor data fetching API docs (#30615)
vercel/next.js: docs/basic-features/data-fetching/getServerSideProps.md

---

> description: Fetch data at build time with `getStaticProps` API reference.

Reviewer: ```suggestion
description: Fetch data on each request with `getServerSideProps`.
```

---

# `getServerSideProps`

If you export an `async` function called `getServerSideProps` (server-side rendering) from a page, Next.js will pre-render this page on each request using the data returned by `getServerSideProps`.

```js
export async function getServerSideProps(context) {
  return {
    props: {}, // will be passed to the page component as props
  }
}
```

The [`getServerSideProps` API reference](/docs/api-reference/data-fetching/getServerSideProps.md) covers all parameters and props that can be used with `getServerSideProps`.

## When should I use `getServerSideProps`?

> You should use `getServerSideProps` only if you need to pre-render a page whose data must be fetched at request time. Time to first byte (TTFB) will be slower than [`getStaticProps`](/docs/basic-features/data-fetching/getStaticProps.md) because the server must compute the result on every request, and the result cannot be cached by a CDN without extra configuration.

Reviewer: ```suggestion
You should use `getServerSideProps` only if you need to pre-render a page whose data must be fetched at request time. [Time to First Byte (TTFB)](/learn/seo/web-performance) will be slower than [`getStaticProps`](/docs/basic-features/data-fetching/getStaticProps.md) because the server must compute the result on every request, and the result can only be cached by a CDN using `cache-control` headers (which could require extra configuration).
```

You can use the [next-code-elimination tool](https://next-code-elimination.vercel.app/) to verify what Next.js eliminates from the client-side bundle.

`getServerSideProps` can only be exported from a **page**. You can’t export it from non-page files.

Also, you must use `export async function getServerSideProps() {}` — it will **not** work if you add `getServerSideProps` as a property of the page component.

## Fetching data on the client side

If your page contains frequently updating data, and you don’t need to pre-render the data, you can fetch the data on the client side. An example of this is user-specific data:

- First, immediately show the page without data. Parts of the page can be pre-rendered using Static Generation. You can show loading states for missing data.
- Then, fetch the data on the client side and display it when ready.

> This approach works well for user dashboard pages, for example. Because a dashboard is a private, user-specific page, SEO is not relevant and the page doesn’t need to be pre-rendered. The data is frequently updated, which requires request-time data fetching.

Reviewer: Could we link to an example?

Reviewer 2: +1

You should use `getServerSideProps` only if you need to pre-render a page whose data must be fetched at request time. [Time to First Byte (TTFB)](/learn/seo/web-performance) will be slower than [`getStaticProps`](/docs/basic-features/data-fetching/getStaticProps.md) because the server must compute the result on every request, and the result can only be cached by a CDN using `cache-control` headers (which could require extra configuration).

If you do not need to pre-render the data, then you should consider fetching data on the [client side](#fetching-data-on-the-client-side).

### `getServerSideProps` or API Routes?

It can be tempting to reach for an [API Route](/docs/api-routes/introduction.md) when you want to fetch data from the server, then call that API route from `getServerSideProps`. This is an unnecessary and inefficient approach, as it will cause an extra request to be made due to both `getServerSideProps` and API Routes running on the server.

Instead, directly import the logic used inside your API Route into `getServerSideProps`. This could mean calling a CMS, database, or other API directly from inside `getServerSideProps`.

> ### TypeScript: Use `GetServerSideProps`

Reviewer 3: these typescript sections feel slightly out of place but i can also understand the struggle to find the right spot for them. could you share your thinking here about the typescript api being more "concept" than "reference"? i am curious (i don't have a strong opinion about which it is)

Author: Oh wow, thats a mess up from me. They should be in the api ref docs. I have getServerSideProps, but didnt move the others. Thank you for spotting this!

For TypeScript, you can use the `GetServerSideProps` type from `next`:

```ts
import { GetServerSideProps } from 'next'

export const getServerSideProps: GetServerSideProps = async (context) => {
  // ...
}
```

If you want to get inferred typings for your props, you can use `InferGetServerSidePropsType<typeof getServerSideProps>`:

```tsx
import { InferGetServerSidePropsType } from 'next'

type Data = { ... }

export const getServerSideProps = async () => {
  const res = await fetch('https://.../data')
  const data: Data = await res.json()

  return {
    props: {
      data,
    },
  }
}

function Page({ data }: InferGetServerSidePropsType<typeof getServerSideProps>) {
  // will resolve posts to type Data
}

export default Page
```

> ### Technical details

Reviewer 3: could you say more about what "technical details" as a title means for you? it seems like something that unless i'm just randomly curious i could skip, but i'm not sure i agree with that about all the content here!

Author: This is a relic from the previous docs that we dont need, the sub headings can be promoted!

#### Only runs on server-side

`getServerSideProps` only runs on server-side and never runs on the browser. If a page uses `getServerSideProps`, then:

- When you request this page directly, `getServerSideProps` runs at the request time, and this page will be pre-rendered with the returned props

> - When you request this page on client-side page transitions through ([`next/link`](/docs/api-reference/next/link.md)) or ([`next/router`](/docs/api-reference/next/router.md)), Next.js sends an API request to the server, which runs `getServerSideProps`. It’ll return JSON that contains the result of running `getServerSideProps`, and the `JSON` will be used to render the page. All this work will be handled automatically by Next.js, so you don’t need to do anything extra as long as you have `getServerSideProps` defined

Reviewer 4: ```suggestion
- When you request this page on client-side page transitions through [`next/link`](/docs/api-reference/next/link.md) or [`next/router`](/docs/api-reference/next/router.md), Next.js sends an API request to the server, which runs `getServerSideProps`. It’ll return JSON that contains the result of running `getServerSideProps`, and the `JSON` will be used to render the page. All this work will be handled automatically by Next.js, so you don’t need to do anything extra as long as you have `getServerSideProps` defined
```

You can use the [next-code-elimination tool](https://next-code-elimination.vercel.app/) to verify what Next.js eliminates from the client-side bundle.

`getServerSideProps` can only be exported from a **page**. You can’t export it from non-page files.

> Also, you must use `export async function getServerSideProps() {}` — it will **not** work if you add `getServerSideProps` as a property of the page component.

Reviewer 4: ```suggestion
Also, you must export `getServerSideProps` as a standalone function. — it will **not** work if you add `getServerSideProps` as a property of the page component.
```

The TypeScript section used the form `export const getServerSideProp...` so it should be confusing to first-time users to understand which form is correct. Both are.
